### PR TITLE
Add back signing

### PR DESCRIPTION
**Pull request by costin-zaharia-sonarsource**
Fixes: #8230

Signing has been disabled by #8229

smctl doc: https://docs.digicert.com/en/digicert-keylocker/tools/command-line-interface/smctl.html

### CODE REVIEW

File: azure-pipelines.yml [hunk 1]
```diff
@@ -10,12 +10,17 @@ schedules:
 trigger:
 - master
 
-pool: '.Net Bubble - GCP'
+# ToDo: Remove this change once the staging image changes are merged in the base image 
+pool:
+  name: '.Net Bubble staging'
+  demands:
+    - PR_NUMBER -equals 381
```
Comment (costin-zaharia-sonarsource) [inline azure-pipelines.yml:17]: I will remove this when https://cirrus-ci.com/build/6625351292420096 is green

File: azure-pipelines.yml [hunk 2]
```diff
@@ -95,24 +102,56 @@ stages:
...
+      - task: DownloadSecureFile@1
+        displayName: 'Download crt file'
+        name: SM_CLIENT_CRT
+        inputs:
+          secureFile: cert_525594307.crt
+
+      - task: PowerShell@2
+        displayName: "Signing certificate setup"
+        # ToDo: add back the condition
+        # condition: eq(variables.isReleaseBranch, 'True')
```
Comment (costin-zaharia-sonarsource) [inline azure-pipelines.yml:119]: I will remove the ToDos once the PR is "ready" to be merged so we can validate the signing until then

File: analyzers/src/SonarAnalyzer.CFG/SonarAnalyzer.CFG.csproj [hunk 3]
```diff
@@ -47,7 +47,7 @@
 
   <Target Name="SignDlls" Condition="'$(SignAssembly)' == 'true' and $(TargetFramework) != ''" AfterTargets="Build">
     <Message Importance="high" Text="Signing SonarAnalyzer.CFG assemblies from $(OutDir)" />
-    <Exec WorkingDirectory="$(OutDir)" Command="&quot;$(SIGNTOOL_PATH)&quot; sign /fd SHA256 /f $(PFX_PATH) /p $(PFX_PASSWORD) /tr http://sha256timestamp.ws.symantec.com/sha256/timestamp SonarAnalyzer.CFG.dll" />
+    <Exec WorkingDirectory="$(OutDir)" Command="&quot;$(SIGNTOOL_PATH)&quot; sign /tr http://timestamp.digicert.com /td SHA256 /fd SHA256 /csp &quot;DigiCert Signing Manager KSP&quot; /kc &quot;$(SM_KP)&quot; /f &quot;$(SM_CLIENT_CRT_FILE)&quot; SonarAnalyzer.CFG.dll" />
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/SonarAnalyzer.CFG.csproj:50]: I think the magic happens with the `/csp` parameter:
> /csp CSPName | Specifies the cryptographic service provider (CSP) that contains the private key container.

[Source](https://learn.microsoft.com/en-us/dotnet/framework/tools/signtool-exe#sign-command-options)

This provider most likely contacts the digi cert server and authenticates via the SM_CLIENT_CERT_FILE file. The server provides the private key and the `DigiCert Signing Manager KSP` hands it over to sign.exe which signs it with the certificate given in `SM_CLIENT_CRT_FILE`

switch | Description | Our value
-- | -- | --
/f SignCertFile | Specifies the signing certificate in a file. If the file is in Personal Information Exchange (PFX) format and protected by a password, use the /p option to specify the password. If the file does not contain private keys, use the /csp and /kc options to specify the CSP and private key container name.  | SM_CLIENT_CRT_FILE (without private key)
/kc PrivKeyContainerName | Specifies the private key container name. | SM_KP
/csp CSPName | Specifies the cryptographic service provider (CSP) that contains the private key container. | "DigiCert Signing Manager KSP"

File: azure-pipelines.yml [hunk 4]
```diff
@@ -95,24 +98,56 @@ stages:
...
+        env:
+          SIGNTOOL_PATH: $(SignToolPath)
+          SM_CLIENT_CERT_FILE: $(SM_CLIENT_CERT.secureFilePath)
+          SM_CLIENT_CERT_PASSWORD: $(SM_CLIENT_CERT_PASSWORD)
+          SM_API_KEY: $(SM_API_KEY)
+          SM_CERT: $(SM_CERT)
+          SM_CERT_FP: $(SM_CERT_FP)
+          SM_HOST: $(SM_HOST)
+          SM_KP: $(SM_KP)
```
Comment (martin-strecker-sonarsource) [inline azure-pipelines.yml:124]: You can remove the values that are not secret in the variable group as they should be available automatically.

File: azure-pipelines.yml [hunk 5]
```diff
@@ -125,15 +160,24 @@ stages:
...
+        inputs:
+          targetType: 'inline'
+          script: |
+            nuget sign "$env:PACKAGES_PATH" -Overwrite -HashAlgorithm SHA256 -CertificateFingerprint $env:SM_CERT_FP -Timestamper http://timestamp.digicert.com -TimestampHashAlgorithm SHA256
```
Comment (martin-strecker-sonarsource) [inline azure-pipelines.yml:180]: There is no need to copy `$env:SM_CERT_FP` to the env and access it here. You can use the variable $(SM_CERT_FP) directly here.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
I think we can simplify a couple of settings. There should be no need to copy non-secret variables to environment variables as this is done automatically. Just make sure the variables are named accordingly (which is the case in mot cases).

File: analyzers/src/SonarAnalyzer.CFG/SonarAnalyzer.CFG.csproj (hunk 3 above)
[the remark this answers is not in the document]
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/SonarAnalyzer.CFG.csproj:50]: I don't know why the signing was implemented this way. 

I've created an issue to extract it: https://github.com/SonarSource/sonar-dotnet/issues/8343

File: analyzers/src/SonarAnalyzer.CSharp/SonarAnalyzer.CSharp.csproj [hunk 6]
```diff
@@ -44,7 +44,7 @@
 
   <Target Name="SignDlls" Condition="'$(SignAssembly)' == 'true'">
     <Message Importance="high" Text="Signing analyzer assemblies from $(OutputPath)" />
-    <Exec WorkingDirectory="$(OutputPath)" Command="&quot;$(SIGNTOOL_PATH)&quot; sign /fd SHA256 /f $(PFX_PATH) /p $(PFX_PASSWORD) /du https://www.sonarsource.com/ /d &quot;Roslyn analyzers that spot Bugs, Vulnerabilities and Code Smells in your code.&quot; /tr http://sha256timestamp.ws.symantec.com/sha256/timestamp SonarAnalyzer*.dll" />
+    <Exec WorkingDirectory="$(OutputPath)" Command="signtool sign /du https://www.sonarsource.com/ /d &quot;Roslyn analyzers that spot Bugs, Vulnerabilities and Code Smells in your code.&quot; /tr http://timestamp.digicert.com /td SHA256 /fd SHA256 /csp &quot;DigiCert Signing Manager KSP&quot; /kc &quot;$(SM_KP)&quot; /f &quot;$(SM_CLIENT_CRT_FILE)&quot; SonarAnalyzer*.dll" />
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/SonarAnalyzer.CSharp.csproj:47]: This signs `SonarAnalyzer*.dll` which likely leads to double or triple signing of SonarAnalyzer.Commons and SonarAnalyzer.CFG (overwriting the previous assignment). I added a https://github.com/SonarSource/sonar-dotnet/issues/8343#issuecomment-1805341576

File: azure-pipelines.yml [hunk 7]
```diff
@@ -95,24 +96,48 @@ stages:
...
+        env:
+          SM_CLIENT_CERT_FILE: $(SM_CLIENT_CERT.secureFilePath)
+          SM_CLIENT_CERT_PASSWORD: $(SM_CLIENT_CERT_PASSWORD)
+          SM_API_KEY: $(SM_API_KEY)
+          SM_CERT: $(SM_CERT)
+        inputs:
+          targetType: 'inline'
+          script: |
+            Write-Output "smctl healthcheck:"
+            smctl healthcheck
```
Comment (tomverin) [inline azure-pipelines.yml:123]: You can remove healthcheck now that everything is working fine

**Review by tomverin — APPROVED** [review]

File: azure-pipelines.yml [hunk 8]
```diff
@@ -95,24 +96,48 @@ stages:
...
+          secureFile: digicert_authentication_certificate.p12
+ 
+      - task: DownloadSecureFile@1
+        displayName: 'Download crt file'
+        name: SM_CLIENT_CRT
+        inputs:
+          secureFile: cert_525594307.crt
+
+      - task: PowerShell@2
+        displayName: "Signing certificate setup"
```
Comment (martin-strecker-sonarsource) [inline azure-pipelines.yml:111]: I think this tasks could benefit from a brief description of what we assume it does:

> Initialize the DigiCert Private Key Provider. What we think it does: The smctl tool authenticates with a client certificate (SM_CLIENT_CERT_FILE) and a client password (SM_CLIENT_CERT_PASSWORD). It uses an API Key (SM_API_KEY) and the ID of the certificate (SM_CERT) to check if the authenticated client is authorized to use the certificate specified and synchronize (potentially private) information about the certificate.

That is a lot of speculation but better than not having any information at all.

File: azure-pipelines.yml [hunk 9]
```diff
@@ -95,24 +96,48 @@ stages:
           secureFile: SonarSourceSecret.snk
 
       - task: DownloadSecureFile@1
-        displayName: 'Download pfx file'
-        name: pfx
+        displayName: 'Download p12 file'
+        name: SM_CLIENT_CERT
         inputs:
```
Comment (martin-strecker-sonarsource) [inline azure-pipelines.yml:101]: Add a comment that this file is used by the `DigiCert Signing Manager KSP` Key Storage Provider to authenticate against the DigiCert private key provider server.

File: azure-pipelines.yml [hunk 10]
```diff
@@ -95,24 +96,48 @@ stages:
...
-          secureFile: 'SonarSource-2021-2023.pfx'
+          secureFile: digicert_authentication_certificate.p12
+ 
+      - task: DownloadSecureFile@1
+        displayName: 'Download crt file'
+        name: SM_CLIENT_CRT
+        inputs:
+          secureFile: cert_525594307.crt
```
Comment (martin-strecker-sonarsource) [inline azure-pipelines.yml:108]: Add a comment that this is the signing certificate without the private key.

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. I would add some comments in the Yaml about what we think the "SM_" entries are used for. It took us quite a while to figure this out; even if we are unsure, it is better to keep the information for our future selves.

File: analyzers/src/SonarAnalyzer.CSharp/SonarAnalyzer.CSharp.csproj (hunk 6 above)
Comment (costin-zaharia-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/SonarAnalyzer.CSharp.csproj:47]: Sounds good. Nothing to do in this PR, right?

**Pull request merged**